### DISCUSSION TITLE
생성자 대신 정적 팩터리 메서드를 고려하라

### DISCUSSION THREAD

**Discussion by shkisme** [BDD-CLUB/challenge-every-week-posting#46] [1. 개발 관련 글 모음]
[생성자 대신 정적 팩터리 메서드를 고려하라](https://velog.io/@shkisme/Item-1-%EC%83%9D%EC%84%B1%EC%9E%90-%EB%8C%80%EC%8B%A0-%EC%A0%95%EC%A0%81-%ED%8C%A9%ED%84%B0%EB%A6%AC-%EB%A9%94%EC%84%9C%EB%93%9C%EB%A5%BC-%EA%B3%A0%EB%A0%A4%ED%95%98%EB%9D%BC)

이번 글은 이펙티브 자바 아이템 1 내용을 읽고 정리한 글입니다. 예전에 휘갈겨 작성한 글을 다듬어서 남에게 설명하듯 최대한 알아보기 쉽게 다시 작성했습니다.
+) 책에 기록되어 있는 예제 코드가 적절하고 알아보기도 쉽다고 느껴서 따로 코드를 만들어보지는 않았습니다!

**Comment by pnudpwls1103** [comment]
혹시 글에서 아래 내용에 대해 더 설명해주실 수 있나요? 잘 이해가 안 가서 여쭤봅니다. ㅎ
- 3번 내용 중 '성능 개선을 위해 하위 타입 클래스를 더 추가하여 릴리즈 해도 상관 없을 것입니다.' 
- 4번 정적 팩터리 메서드를 작성하는 시점에는 반환할 객체의 클래스가 존재하지 않아도 된다.

**Reply by shkisme** [reply]
3번의 예제 코드로 아래 코드를 들었습니다!
```java
public static <E extends Enum<E>> EnumSet<E> noneOf(Class<E> elementType) {
    Enum<?>[] universe = getUniverse(elementType);
    if (universe == null)
        throw new ClassCastException(elementType + " not an enum");

    if (universe.length <= 64)
        return new RegularEnumSet<>(elementType, universe);
    else
        return new JumboEnumSet<>(elementType, universe);
}
```
코드를 보면 파라미터에 따라서 universe.length값이 달라지고, 이 값을 기준으로 값이 작거나 같다면 RegularEnumSet, 크다면 JumboEnumSet 로 다른 인스턴스를 반환하고 있습니다. (EnumSet은 abstract 클래스이고 RegularEnumSet, JumboEnumSet 두 클래스는 EnumSet을 상속받는 클래스 입니다.)

그렇다면 클라이언트는 아래 처럼 정적 팩터리 메서드를 호출할 것입니다!
```java
EnumSet enumSet = EnumSet.noneOf(elementType);
```

여기서 성능 개선을 위해 하위 타입 클래스 GiantEnumSet를 추가해보겠습니다. GiantEnumSet도 마찬가지로 EnumSet을 상속받겠습니다.

```java
public static <E extends Enum<E>> EnumSet<E> noneOf(Class<E> elementType) {
    ...
    if (universe.length <= 64)
        return new RegularEnumSet<>(elementType, universe);
    else if (universe.length <= 128)
        return new JumboEnumSet<>(elementType, universe);
    else
        return new GiantEnumSet<>(elementType, universe);
}
```

위 처럼 api를 개선하여 릴리즈 한다 해도, 클라이언트가 예전에 작성한 코드에는 문제가 없어야 합니다. 또한 클라이언트가 앞으로 작성할 코드 역시 변함 없이 아래와 같을 겁니다!

```java
EnumSet enumSet = EnumSet.noneOf(elementType);
```

---

4번의 경우에는 TODO로 남겨둔 만큼, 아직 저 역시 JDBC에 대한 이해가 부족하여 조사중이었는데요 ㅎㅎ 
[참고한 블로그 글](https://sysgongbu.tistory.com/95)의 예시 코드를 바탕으로 설명드리겠습니다!

```java
public static void main(String[] args) {
	String driverName = "com.mysql.jdbc.Driver";
	String url = "jdbc:mysql://localhost:3306/test";
	String user = "root";
	String password = "1234";
    
	try {
		Class.forName(driverName); 

		// 서비스 접근 API인 DriverManager.getConnection가 서비스 구현체(서비스 인터페이스)인 Connection 반환
		Connection connection = DriverManager.getConnection(url, user, password);
	} catch (ClassNotFoundException e) {
		e.printStackTrace();
	} catch (SQLException e) {
		e.printStackTrace();
	}
}
```

- Class.forName(driverName); 에서 문자열로 전달되는 "com.mysql.jdbc.Driver"이라는 클래스가 메모리에 로드 된다.
  - 제 생각에는 바로 이 부분이 정적 팩터리 메서드를 작성하는 시점에는 반환할 객체의 클래스가 존재하지 않아도 된다. 는 말인 것 같습니다!
  - 어떤 데이터 베이스를 사용할 지 모르기 때문에, JDBC는 동적으로 클래스를 가져오는 방식을 택해야 했고 그런 설계를 정적 팩터리 메서드의 특징을 활용하여 구현한 것 같습니다!
  - 만약 mysql이 아닌 다른 데이터 베이스를 사용한다 해도, "com.mysql.jdbc.Driver"를 바꿔 끼우면 되는 유연함을 가질 수 있습니다.
  - 이는 DI(의존성 주입) 등의 설계에도 근간이 되었다고 합니다!
 
 ---

제 글에서 설명이 미흡한 부분을 명확히 찾을 수 있는 질문 굿입니다!! 👍👍

**Reply by pnudpwls1103** [reply]
오오 확실히 이해가 되는거 같네요!
그 4번에서는 Class 클래스에는 아래와 같이 미리 인스턴스를 만들어 놓고 문자열에 따라 반환한다고 이해하면 되는거 맞죠?!
(Class.java)
public static final (DB 클래스) db1 = new (DB 클래스)();
public static final (DB 클래스) db2 = new (DB 클래스)();

-------------------------------------------------
근데 정적 팩토리 메서드에 static final 로 생성한 인스턴스가 너무 많으면 메모리에 안 쓰는데 할당된 인스턴스가 공간을 차지하는 문제도 있지 않을까하는 생각이 갑자기 떠오르네요 🤔

**Reply by shkisme** [reply]
> 그 4번에서는 Class 클래스에는 아래와 같이 미리 인스턴스를 만들어 놓고 문자열에 따라 반환한다고 이해하면 되는거 맞죠?!
(Class.java)
public static final (DB 클래스) db1 = new (DB 클래스)();
public static final (DB 클래스) db2 = new (DB 클래스)();

아마도 클래스를 동적으로 가져오는 데는 위 방식 보다는 Reflecton api를 사용할 것 같은데 내부 동작 원리는 저도 자세히 모르겠네요...! 이것도 한번 알아봐야겠습니다 ㅎㅎ

> 근데 정적 팩토리 메서드에 static final 로 생성한 인스턴스가 너무 많으면 메모리에 안 쓰는데 할당된 인스턴스가 공간을 차지하는 문제도 있지 않을까하는 생각이 갑자기 떠오르네요 🤔

그렇기에 주로 비싼 인스턴스 & 재사용성이 높은 걸 static final 로 생성해두고 그걸 재사용하는 방식을 택하는 것 같아요! 그리고 정적 팩터리로 인한 오버헤드에 관한 이야기는 거의 없는 것 같아서 그 영향이 미미하다고 생각하고 있습니다!

**Reply by pnudpwls1103** [reply]
친절한 설명 너무너무 감사합니다!!
